File: src/main/java/factorio/debugger/DAP/DAPMessageParser.java
package factorio.debugger.DAP;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import org.jetbrains.annotations.NotNull;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.ObjectWriter;
import com.intellij.openapi.diagnostic.Logger;
import factorio.debugger.DAP.messages.DAPEvent;
import factorio.debugger.DAP.messages.DAPProtocolMessage;
import factorio.debugger.DAP.messages.DAPRequest;
import factorio.debugger.DAP.messages.DAPResponse;

public class DAPMessageParser {
    private final Logger logger = Logger.getInstance(DAPMessageParser.class);
    public static final String contentLengthHeader = "Content-Length: ";

    private String myLeftoverBuffer;
    private boolean headerComplete;
    private int myContentLength;
    private final ObjectMapper myObjectMapper;

    private @NotNull ArrayList<DAPProtocolMessage> myDapMessages;

    public DAPMessageParser() {
        myObjectMapper = new ObjectMapper();
        myDapMessages = new ArrayList<>();
    }

    protected ObjectWriter getPrettyPrinter() {
        return myObjectMapper.writerWithDefaultPrettyPrinter();
    }

    protected List<DAPProtocolMessage> getParsedMessages() {
        if(myDapMessages.isEmpty()) return Collections.emptyList();

        List<DAPProtocolMessage> result = myDapMessages;
        myDapMessages = new ArrayList<>();
        return result;
    }

    public List<DAPProtocolMessage> parse(@NotNull final String text) {
        String input = myLeftoverBuffer + text;

        int offset = 0, lastOffset = 0;
        while(offset >= 0) {
            lastOffset = offset;
            offset = parseMessage(input, offset);
        }

        if (lastOffset < input.length()) {
            // we have left over characters
            myLeftoverBuffer = input.substring(lastOffset);
        } else {
            myLeftoverBuffer = "";
        }

        return getParsedMessages();
    }

    private int parseMessage(final String input, final int offset) {
        int messageStart = parseMessageHeader(input, offset);
        if (messageStart < 0) return -1;

        if (!headerComplete) {
            // error while parsing the header
            if (messageStart > offset) {
                // skip forward
                return messageStart;
            } else {
                // not enough characters to parse header
                // add header to leftovers
                return -1;
            }
        }

        // header parsed successfully
        return parseMessageContent(input, messageStart);
    }

    private int parseMessageContent(final String input, final int messageStart) {
        /*if (input.length() < messageStart + myContentLength) {
            // input doesn't contain the whole message
            return -1;
        }*/

        int messageEnd = Math.min(input.length(), messageStart + myContentLength);

        String myJsonMessage = input.substring(messageStart, messageEnd);

        try {
            JsonNode node = myObjectMapper.readTree(myJsonMessage);

            String messageType = node.at("/type").asText("");

            DAPProtocolMessage message = switch (messageType) {
                case "request" -> myObjectMapper.treeToValue(node, DAPRequest.class);
                case "response" -> myObjectMapper.treeToValue(node, DAPResponse.class);
                case "event" -> myObjectMapper.treeToValue(node, DAPEvent.class);
                default -> null;
            };

            if (message != null) {
                myDapMessages.add(message);
            } else {
                logger.info(String.format("Ignored message: %s", getPrettyPrinter().writeValueAsString(node)));
            }
        } catch (JsonProcessingException e) {
            if (input.length() < messageStart + myContentLength) {
                // input doesn't contain the whole message
                return -1;
            }

            logger.warn("Received json: "+myJsonMessage);
            logger.warn("Unable to parse message!", e);
        }

        return messageEnd;
    }

    private int parseMessageHeader(@NotNull final String text, int offset) {
        int headerStart = text.indexOf(contentLengthHeader, offset);
        headerComplete = false;
        if (headerStart < 0) return -1;

        int endOfLine = text.indexOf("\r\n", headerStart);

        // not enough characters to fully parse the header yet
        if(endOfLine < 0 || text.length() < endOfLine + 4) return -1;

        String contentLengthStr = text.substring(headerStart + contentLengthHeader.length(), endOfLine);
        if(text.regionMatches(endOfLine, "\r\n\r\n", 0, 4)) {
            try {
                int contentLength = Integer.parseInt(contentLengthStr);

                //logger.warn("Expecting message of size: "+contentLength);
                if (contentLength < 10485760) { // 10 MB content length max
                    headerComplete = true;
                    myContentLength = contentLength;
                    return endOfLine+4;
                } else {
                    logger.warn("Content length too big! "+contentLength);
                }
            } catch (NumberFormatException e) {
                logger.warn("Invalid content length: '"+contentLengthStr+"'");
            }
        } else {
            logger.warn(String.format("Invalid DAP header: No new lines int \"%s\"", text.substring(headerStart, endOfLine+4)));
        }
        return endOfLine;
    }
}

File: src/main/kotlin/factorio/debugger/DAP/DAPMessageParser.kt
package factorio.debugger.DAP

import com.fasterxml.jackson.core.JsonProcessingException
import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.databind.ObjectWriter
import com.intellij.openapi.diagnostic.Logger
import factorio.debugger.DAP.messages.DAPProtocolMessage
import factorio.debugger.DAP.messages.events.DAPEvent
import factorio.debugger.DAP.messages.requests.DAPRequest
import factorio.debugger.DAP.messages.responses.DAPResponse

open class DAPMessageParser {
    private val logger = Logger.getInstance(DAPMessageParser::class.java)
    private var myLeftoverBuffer: String? = null
    private var headerComplete = false
    private var myContentLength = 0
    private val myObjectMapper: ObjectMapper = ObjectMapper()
    private var myDapMessages: ArrayList<DAPProtocolMessage> = ArrayList()

    protected val prettyPrinter: ObjectWriter
        get() = myObjectMapper.writerWithDefaultPrettyPrinter()
    protected val parsedMessages: List<DAPProtocolMessage>
        get() {
            if (myDapMessages.isEmpty()) return emptyList()
            val result: List<DAPProtocolMessage> = myDapMessages
            myDapMessages = ArrayList()
            return result
        }

    fun parse(text: String): List<DAPProtocolMessage> {
        val input = myLeftoverBuffer + text
        var offset = 0
        var lastOffset = 0
        while (offset >= 0) {
            lastOffset = offset
            offset = parseMessage(input, offset)
        }
        myLeftoverBuffer = if (lastOffset < input.length) {
            // we have left over characters
            input.substring(lastOffset)
        } else {
            ""
        }
        return parsedMessages
    }

    private fun parseMessage(input: String, offset: Int): Int {
        val messageStart = parseMessageHeader(input, offset)
        if (messageStart < 0) return -1
        return if (!headerComplete) {
            // error while parsing the header
            if (messageStart > offset) {
                // skip forward
                messageStart
            } else {
                // not enough characters to parse header
                // add header to leftovers
                -1
            }
        } else parseMessageContent(input, messageStart)

        // header parsed successfully
    }

    private fun parseMessageContent(input: String, messageStart: Int): Int {
        /*if (input.length() < messageStart + myContentLength) {
            // input doesn't contain the whole message
            return -1;
        }*/
        val messageEnd = input.length.coerceAtMost(messageStart + myContentLength)
        val myJsonMessage = input.substring(messageStart, messageEnd)
        try {
            val node = myObjectMapper.readTree(myJsonMessage)
            val message = when (node.at("/type").asText("")) {
                "request" -> myObjectMapper.treeToValue(node, DAPRequest::class.java)
                "response" -> myObjectMapper.treeToValue(node, DAPResponse::class.java)
                "event" -> myObjectMapper.treeToValue(node, DAPEvent::class.java)
                else -> null
            }
            if (message != null) {
                myDapMessages.add(message)
            } else {
                logger.info("Ignored message: ${prettyPrinter.writeValueAsString(node)}")
            }
        } catch (e: JsonProcessingException) {
            if (input.length < messageStart + myContentLength) {
                // input doesn't contain the whole message
                return -1
            }
            logger.warn("Received json: $myJsonMessage")
            logger.warn("Unable to parse message!", e)
        }
        return messageEnd
    }

    private fun parseMessageHeader(text: String, offset: Int): Int {
        val headerStart = text.indexOf(contentLengthHeader, offset)
        headerComplete = false
        if (headerStart < 0) return -1
        val endOfLine = text.indexOf("\r\n", headerStart)

        // not enough characters to fully parse the header yet
        if (endOfLine < 0 || text.length < endOfLine + 4) return -1
        val contentLengthStr = text.substring(headerStart + contentLengthHeader.length, endOfLine)
        if (text.regionMatches(endOfLine, "\r\n\r\n", 0, 4)) {
            try {
                val contentLength = contentLengthStr.toInt()

                //logger.warn("Expecting message of size: "+contentLength);
                if (contentLength < 10485760) { // 10 MB content length max
                    headerComplete = true
                    myContentLength = contentLength
                    return endOfLine + 4
                } else {
                    logger.warn("Content length too big! $contentLength")
                }
            } catch (e: NumberFormatException) {
                logger.warn("Invalid content length: '$contentLengthStr'")
            }
        } else {
            logger.warn("Invalid DAP header: No new lines int \"${text.substring(headerStart, endOfLine + 4)}\"")
        }
        return endOfLine
    }

    companion object {
        const val contentLengthHeader = "Content-Length: "
    }
}
